src/Microsoft.Data.SqlClient/add-ons/AzureKeyVaultProvider/AzureSqlKeyCryptographer.cs

```diff
@@ -0,0 +1,233 @@
+        /// key into the key dictionary.
+        /// </summary>
+        private readonly ConcurrentDictionary<string, Task<Azure.Response<KeyVaultKey>>> _keyFetchTaskDictionary = new ConcurrentDictionary<string, Task<Azure.Response<KeyVaultKey>>>();
+
+        /// <summary>
+        /// Holds references to the Azure Key Vault keys and maps them to their corresponding Azure Key Vault Key Identifier (URI).
+        /// </summary>
+        private readonly ConcurrentDictionary<string, KeyVaultKey> _keyDictionary = new ConcurrentDictionary<string, KeyVaultKey>();
+
+        /// <summary>
+        /// Holds references to the Azure Key Vault CryptographyClient objects and maps them to their corresponding Azure Key Vault Key Identifier (URI).
+        /// </summary>
+        private readonly ConcurrentDictionary<string, CryptographyClient> cryptoClientDictionary = new ConcurrentDictionary<string, CryptographyClient>();
+
+        /// <summary>
+        /// Constructs a new KeyCryptographer
+        /// </summary>
+        /// <param name="tokenCredential"></param>
+        internal AzureSqlKeyCryptographer(TokenCredential tokenCredential)
+        {
+            TokenCredential = tokenCredential;
+        }
+
+        internal AzureSqlKeyCryptographer(AuthenticationCallback authenticationCallback)
+        {
+            AuthenticationCallback = authenticationCallback;
+            isUsingLegacyAuthentication = true;
+        }
+
+        /// <summary>
+        /// Adds the key, specified by the Key Identifier URI, to the cache.
+        /// </summary>
+        /// <param name="keyIdentifierUri"></param>
+        internal void AddKey(string keyIdentifierUri)
+        {
+            if (TheKeyHasNotBeenCached(keyIdentifierUri))
+            {
+                if (isUsingLegacyAuthentication)
+                {
+                    TokenCredential = new AzureKeyVaultProviderTokenCredential(AuthenticationCallback, keyIdentifierUri);
+                }
+
+                ParseAKVPath(keyIdentifierUri, out Uri vaultUri, out string keyName);
+                CreateKeyClient(vaultUri);
+                FetchKey(vaultUri, keyName, keyIdentifierUri);
+            }
+
+            bool TheKeyHasNotBeenCached(string k) => !_keyDictionary.ContainsKey(k) && !_keyFetchTaskDictionary.ContainsKey(k);
+        }
+
+        /// <summary>
+        /// Returns the key specified by the Key Identifier URI
+        /// </summary>
+        /// <param name="keyIdentifierUri"></param>
+        /// <returns></returns>
+        internal KeyVaultKey GetKey(string keyIdentifierUri)
+        {
+            if (_keyDictionary.ContainsKey(keyIdentifierUri))
+            {
+                return _keyDictionary[keyIdentifierUri];
+            }
+
+            if (_keyFetchTaskDictionary.ContainsKey(keyIdentifierUri))
+            {
+                return Task.Run(() => _keyFetchTaskDictionary[keyIdentifierUri]).Result;
```

Comment:
`Result` can result in deadlocks. As much as possible, use async code. Where you can't, use `GetAwaiter().GetResult()`.